agent/xds/routes.go

```diff
@@ -2,18 +2,301 @@ package xds
+
+	return resources, nil
+}
+
+func makeUpstreamRouteForDiscoveryChain(
+	u *structs.Upstream,
+	chain *structs.CompiledDiscoveryChain,
+	cfgSnap *proxycfg.ConfigSnapshot,
+) (*envoy.RouteConfiguration, error) {
+	upstreamID := u.Identifier()
+	routeName := upstreamID
+
+	// TODO(rb): router configs
+	// TODO(rb): splitter configs
+
+	var routes []envoyroute.Route
+
+	switch chain.Node.Type {
+	case structs.DiscoveryGraphNodeTypeRouter:
+		routes = make([]envoyroute.Route, 0, len(chain.Node.Routes))
+
+		for _, discoveryRoute := range chain.Node.Routes {
+			routeMatch := makeRouteMatchForDiscoveryRoute(discoveryRoute, chain.Protocol)
+
+			// TODO(rb): handle PrefixRewrite
+			// TODO(rb): handle RequestTimeout
+			// TODO(rb): handle Retries
+
+			var (
+				routeAction *envoyroute.Route_Route
+				err         error
+			)
+
+			next := discoveryRoute.DestinationNode
+			if next.Type == structs.DiscoveryGraphNodeTypeSplitter {
+				routeAction, err = makeRouteActionForSplitter(upstreamID, cfgSnap.Datacenter, next.Splits)
+				if err != nil {
+					return nil, err
+				}
+
+			} else if next.Type == structs.DiscoveryGraphNodeTypeGroupResolver {
+				groupResolver := next.GroupResolver
+				routeAction = makeRouteActionForSingleCluster(upstreamID, cfgSnap.Datacenter, groupResolver.Target)
+
+			} else {
+				return nil, fmt.Errorf("unexpected graph node after route %q", next.Type)
+			}
+
+			routes = append(routes, envoyroute.Route{
+				Match:  routeMatch,
+				Action: routeAction,
+			})
+		}
+
+	case structs.DiscoveryGraphNodeTypeSplitter:
+		routeAction, err := makeRouteActionForSplitter(upstreamID, cfgSnap.Datacenter, chain.Node.Splits)
+		if err != nil {
+			return nil, err
+		}
+
+		defaultRoute := envoyroute.Route{
+			Match:  makeDefaultRouteMatch(),
+			Action: routeAction,
+		}
+
+		routes = []envoyroute.Route{defaultRoute}
+
+	case structs.DiscoveryGraphNodeTypeGroupResolver:
+		groupResolver := chain.Node.GroupResolver
+
+		routeAction := makeRouteActionForSingleCluster(upstreamID, cfgSnap.Datacenter, groupResolver.Target)
+
+		defaultRoute := envoyroute.Route{
+			Match:  makeDefaultRouteMatch(),
+			Action: routeAction,
+		}
+
+		routes = []envoyroute.Route{defaultRoute}
+
+	default:
+		panic("unknown top node in discovery chain of type: " + chain.Node.Type)
+	}
+
+	return &envoy.RouteConfiguration{
+		Name: routeName,
+		VirtualHosts: []envoyroute.VirtualHost{
+			envoyroute.VirtualHost{
+				Name:    routeName,
+				Domains: []string{"*"},
+				Routes:  routes,
+			},
+		},
+		// ValidateClusters defaults to true when defined statically and false
+		// when done via RDS. Re-set the sane value of true to prevent
+		// null-routing traffic.
+		ValidateClusters: makeBoolValue(true),
+	}, nil
+}
+
+func makeRouteMatchForDiscoveryRoute(discoveryRoute *structs.DiscoveryRoute, protocol string) envoyroute.RouteMatch {
+	switch protocol {
+	case "http", "http2":
+		// The only match stanza is HTTP.
```

Comment:
Think this is fine. Technically `grpc` is also HTTP2 so we could allow that and then be able to setup per-RPC routing but I'm happy to punt that until we decide if we really want to expose it that way or make a more first-class routing syntax for gRPC method names?

Reply:
Yeah the actual exposed surface area for GA is something we should have a chat about after we can play with the whole thing in practice.